source/link/fnvar.cpp

```diff
@@ -0,0 +1,973 @@
+        const auto& variant = _variants[i];
+        assert(!variant.IsBase());
+
+        for (const auto& fn : *variant.GetModule()) {
+          const auto fn_name = GetFnName(fn.DefInst());
+          if (fn_name == called_fn_name) {
+            called_fns.push_back(std::make_pair(variant.GetName(), &fn));
+          }
+        }
+      }
+
+      if (!called_fns.empty()) {
+        _base_fn_calls[inst->result_id()] = called_fns;
+      }
+    }
+  });
+}
+
+void Variants::CombineBaseFnCalls(IRContext* linked_context) {
+  for (auto kv : _base_fn_calls) {
+    const uint32_t call_id = kv.first;
+    const auto called_fns = kv.second;
+
+    if (called_fns.empty()) {
+      return;
+    }
+
+    opt::BasicBlock* fn_call_bb = linked_context->get_instr_block(call_id);
+
+    Instruction* found_call_inst = nullptr;
+    auto bb_iter = fn_call_bb->begin();
+    while (bb_iter != fn_call_bb->end() && found_call_inst == nullptr) {
+      if (bb_iter->HasResultId() && bb_iter->result_id() == call_id) {
+        found_call_inst = &*bb_iter;
+      }
+      ++bb_iter;
+    }
+
+    if (found_call_inst == nullptr) {
+      return;
+    }
+
+    const auto base_spec_const_id =
+        FindSpecConstByName(_variants[0].GetModule(), _variants[0].GetName());
+    const auto base_type_op = found_call_inst->context()
+                                  ->get_def_use_mgr()
+                                  ->GetDef(found_call_inst->type_id())
+                                  ->opcode();
+    const auto base_call_id = found_call_inst->result_id();
+
+    // decorate the base call with ConditionalINTEL
+    DecorateConditional(linked_context, base_call_id, base_spec_const_id);
+
+    // Add OpFunctionCall for each variant
+    Instruction* last_inst = found_call_inst;
+    std::vector<std::pair<uint32_t, uint32_t>> var_call_ids;
+    for (const auto& kv2 : called_fns) {
+      const std::string var_name = kv2.first;
+      const opt::Function* fn = kv2.second;
+      const uint32_t spec_const_id =
+          FindSpecConstByName(linked_context->module(), var_name);
+      assert(spec_const_id != 0);
+      const uint32_t var_call_id = linked_context->TakeNextId();
+      var_call_ids.push_back(std::make_pair(spec_const_id, var_call_id));
+
+      auto* var_call_inst = found_call_inst->Clone(linked_context);
+      var_call_inst->SetResultId(var_call_id);
+      var_call_inst->SetOperand(2, {fn->result_id()});
+      var_call_inst->InsertAfter(last_inst);
+      linked_context->set_instr_block(var_call_inst, fn_call_bb);
+      last_inst = var_call_inst;
+
+      // decorate the variant call with ConditionalINTEL
+      DecorateConditional(linked_context, var_call_id, spec_const_id);
+    }
+
+    if (base_type_op != spv::Op::OpTypeVoid) {
+      // Add OpConditionalCopyObjectINTEL combining the function calls
+      const uint32_t result_id = linked_context->TakeNextId();
+      auto conditional_copy_inst = last_inst->Clone(linked_context);
```

Comment:
What is the point of cloning this instruction when you seem to be overwriting every operands?

You could just create a new instruction directly:

```
      MakeUnique<Instruction>(context, spv::Op::OpConditionalCopyObjectINTEL, found_call_inst->type_id(), result_id,
                              std::initializer_list<Operand>{}));
```